server/src/main/java/org/elasticsearch/action/bulk/BulkRequest.java

```diff
@@ -154,6 +163,8 @@ public BulkRequest add(IndexRequest request, @Nullable Object payload) {
 
     BulkRequest internalAdd(IndexRequest request, @Nullable Object payload) {
         Objects.requireNonNull(request, "'request' must not be null");
+        applyGlobalParameters(request);
```

Comment:
re 1. agree - will look for usages and make it consistent do use global instead of default.
re 2. I will modify a testcase to show that pipeline and routing can be applied later. Pipeline and routing are not mandatory parameters and are not validated - so should not seem to be immutable. These are applied in RequestConverters so should be possible to modify and apply at the end (works for pipeline now, not routing at the moment - good catch)